src/pyhf/contrib/viz/brazil.py

```diff
@@ -1,9 +1,76 @@
 """Brazil Band Plots."""
 import numpy as np
+from matplotlib.container import Container
 
 
-def plot_results(ax, mutests, tests, test_size=0.05):
+class BrazilBandContainer(Container):
+    r"""
+    Container for the :obj:`matplotlib.artist` objects created in a
+    :func:`~pyhf.contrib.viz.brazil.plot_results` plot.
+
+    The container can be treated like a :obj:`collections.namedtuple`
+    ``(cls_obs, cls_exp, one_sigma_band, two_sigma_band, test_size)``.
+
+    Attributes:
+        cls_obs (:class:`matplotlib.lines.Line2D`): The artist of the
+         :math:`\mathrm{CL}_{s,\mathrm{obs}}` line.
+
+        cls_exp (:obj:`list` of :class:`matplotlib.lines.Line2D`): The artists of
+         the :math:`\mathrm{CL}_{s,\mathrm{exp}}` lines.
+
+        one_sigma_band (:class:`matplotlib.collections.PolyCollection`):
+         The artists of the :math:`\mathrm{CL}_{s,\mathrm{exp}}` :math:`\pm1\sigma`
+         band.
+
+        two_sigma_band (:class:`matplotlib.collections.PolyCollection`):
+         The artists of the :math:`\mathrm{CL}_{s,\mathrm{exp}}` :math:`\pm2\sigma`
+         band.
+
+        test_size (:class:`matplotlib.lines.Line2D`): The artist of the test size
+         line.
+
+        axes (:class:`matplotlib.axes.Axes`): The :class:`matplotlib.axes.Axes`
+         instance the container artists reside in.
     """
+
+    def __init__(self, brazil_band_artists, **kwargs):
+        r"""
+        Args:
+            brazil_band_artists (:obj:`tuple`): Tuple of
+             ``(cls_obs, cls_exp, one_sigma_band, two_sigma_band, test_size)``.
+
+              * ``cls_obs`` contains the :class:`matplotlib.lines.Line2D` of the
+                observed :math:`\mathrm{CL}_{s}` line.
+
+              * ``cls_exp`` is a :obj:`list` of :class:`matplotlib.lines.Line2D` of
+                the expected :math:`\mathrm{CL}_{s}` lines.
+
+              * ``one_sigma_band`` contains the :class:`matplotlib.collections.PolyCollection`
+                of the :math:`\mathrm{CL}_{s,\mathrm{exp}}` :math:`\pm1\sigma` bands.
+
+              * ``two_sigma_band`` contains the :class:`matplotlib.collections.PolyCollection`
+                of the :math:`\mathrm{CL}_{s,\mathrm{exp}}` :math:`\pm2\sigma` bands.
+
+              * ``test_size`` contains the :class:`matplotlib.lines.Line2D` of the test size.
+        """
+        (
+            cls_obs,
+            cls_exp,
+            one_sigma_band,
+            two_sigma_band,
+            test_size,
+        ) = brazil_band_artists
+        self.cls_obs = cls_obs
+        self.cls_exp = cls_exp
+        self.one_sigma_band = one_sigma_band
+        self.two_sigma_band = two_sigma_band
+        self.test_size = test_size
+        self.axes = cls_obs.axes
```

Comment:
I don't recommend this. It's providing a "cute" shortcut but why not just teach users to use `self.cls_obs.axes`? That will work throughout all matplotlib, not just here. If you really need it, or if you can find an example in matplotlib where this is done this way, then it should be a property on BrazilBandContainer, and not stored here.

Reply:
the reason I don't like `artist.cls_obs.axes` is that it provides many ways to do the same thing without a clear canonical choice

e.g. why `artist.cls_obs.axes` and not `artist. one_sigma_band.axes` or `artist.cls_exp.axes`.. providing a canonical way to me seems cleaner and doesn't require knowing/exposing yourself/depending on the details of `BrazilBandContainer`

consider the simple case where I want to to a log scale which has nothinng to do with cls_obs at all

```
artists = make_plot(...)
artists.cls_obs.axes.set_yscale('log')
```

vs

```
artists = make_plot(...)
artists.axes.set_yscale('log')
```

Reply:
Well, if you want more than the basics, I'd probably write:

```python
fig, ax = plt.subplots()
artists = make_plot(..., axes=ax)
ax.set_yscale('log')
```

Now you have access to the figure and the axes, and you can change the size and such when you construct it.

If you are guaranteed that all the Artists are plotted on one Axes only, then I'd at least make this a property.

Reply:
yeah having it as a immutable @property makes sense to me (indeed it's guaranteed for all of those to be onn the same Axes object). we already have the `ax` argument pattern so maybe this is superfluous.. i'll let @matthewfeickert make the call

Reply:
> I don't recommend this. It's providing a "cute" shortcut but why not just teach users to use `self.cls_obs.axes`?

This was my first thought as well. https://github.com/scikit-hep/pyhf/pull/1377#discussion_r612685478 Someone should correct me if I'm wrong, but this (going back to how it was pre adding `axis` as an attribute) also seems in line with the idea that plotting things on an axis in `maptlotlib` returns the artists and you probably are going to be wanting to be explicitly manipulating an `axis` object that already existed instead of the `axis` object that you can get from the artists. 

That is (hopefully more clear in code) this use of the API

```python
fig, ax = plt.subplots()
artists = pyhf.contrib.viz.brazil.plot_results(ax, poi_vals, results)
ax.set_title("API demonstration")
```

is more "natural" in `matplotlib` thinking than the following

```python
fig, ax = plt.subplots()
artists = pyhf.contrib.viz.brazil.plot_results(ax, poi_vals, results)
artists.cls_obs.axes.set_title("API demonstration")  # just use the axis object you already have available to you
```

I'll revert the changes I made in https://github.com/scikit-hep/pyhf/commit/a3e8ec6206e376df5e6e8566b4f3312f22f3a14b once I finish cleaning up the the API changes and tests to move all of the component plotting into kwargs for `plot_results`.

Reply:
That's weird, the ax is a required (non-keyword) argument, in fact the first one? Usually, it's an optional keyword-only argument. Such as in pandas, for example. If you don't have one, it pulls from `plt.gca`, which creates an axis and figure as needed. Not providing `ax=` is generally seen as a shortcut, so "easy" access to the axes is not provided, only Artists.

Reply:
These  seems natural to me

```
fig, ax = plt.subplots()
container = pyhf.contrib.viz.brazil.plot_results(poi_vals, results, ax = ax) #hypoothetical optional kwarg
ax.set_title("API demonstration")
```

```
fig, ax = plt.subplots()
container = pyhf.contrib.viz.brazil.plot_results(poi_vals, results,) #pull from plt.gca()
container.axes.set_title("API demonstration")
```

```
container = pyhf.contrib.viz.brazil.plot_results(poi_vals, results,) #pull from plt.gca()
plt.gca().set_title("API demonstration")
```

while this seems to depend on unnecessary detail

```
fig, ax = plt.subplots()
container = pyhf.contrib.viz.brazil.plot_results(poi_vals, results,) #pull from plt.gca()
container.cls_obs.axes.set_title("API demonstration")
```

but I guess the point then is to generally not access axes from artists.. but either use `plt.gca()` or a pre-created `ax` you anyway have control over

Reply:
the reason `ax` is a non-defaullt argument is that I have a distaste for importing `plt` / relying on global variables like `plt.gcf()`/`plt.gca()` .. but that's probably just prejudice.. (having `ax` kwarg of course important to be able to plot on subplots)

Reply:
1 and 3 are natural. Why do you expect the thing that `pyhf.contrib.viz.brazil.plot_results` returns to have a `.axes` property? Many matplotlib functions return lists of Axis, so calling `.axes` on the return value is not something you can always do - it depends on the return type. While the other two will always work. You should mostly consider this a return tuple of axes, that happens to be "nicer" and provide named access. If you provide `.axes`, why not `.figure`? Where does it stop?

And, currently, you _have_ to pass in axis, as the first argument (?), so you can't do anything besides the first example (and not written that way).

Reply:
(I'm not horribly against providing the property if you are sure they will always be plotted on the same axis, just arguing from a purity/API simplicity/teaching standpoint)

Reply:
yeah i think this PR does double duty in us rethinking how we want to do plotting, so part of that would be to allow a default `ax` value.   The reason I find it natural to be able to call `.axes`on the return value of a complex plotting function like the above  is that it often works for simplpe plotting functions


```
a = plt.scatter([1,2],[2,3])
a.axes.set_yscale('log')
```

```
a = plt.plot([1,2],[2,3])
a[0].axes.set_yscale('log')
```

so extending from "simple plot" to "complex plot" it seems nice to keep that pattern

Reply:
After looking at the docs I _think_ the API argument for no `axes` attribute on containers is the following:

[`scatter()`](https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.scatter.html) returns a [`matplotlib.collections.PathCollection`](https://matplotlib.org/stable/api/collections_api.html#matplotlib-collections) 

```python
>>> import matplotlib.pyplot as plt
>>> fig, ax = plt.subplots()
>>> _scatter = ax.scatter([1, 2], [2, 3])
>>> type(_scatter)
<class 'matplotlib.collections.PathCollection'>
```

which I see from the inheritance diagram is descended from `Artist`.

[![inheritance](https://matplotlib.org/stable/_images/inheritance-6754a4749e8fcb4c370eb0cc2ce674eed0f50813.png)](https://matplotlib.org/stable/api/collections_api.html#matplotlib-collections)

So as all `artists` have `axes` attributes, then so does `collection`. [`plot()`](https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.plot.html) returns a list of [`matplotlib.lines.Line2D`](https://matplotlib.org/stable/api/_as_gen/matplotlib.lines.Line2D.html#matplotlib.lines.Line2D) artists.

```python
>>> import matplotlib.pyplot as plt
>>> fig, ax = plt.subplots()
>>> _plot = ax.plot([1, 2], [2, 3])
>>> _plot
[<matplotlib.lines.Line2D object at 0x7f8dcd04dbe0>]
```

I don't understand _why_ `scatter` doesn't return a collection, but that's I guess a different story.

I guess my point from all that rambling is that both the returns of `scatter` and `plot` don't return [`matplotlib.container`s](https://matplotlib.org/stable/api/container_api.html), which don't have `axes` as they are groupings of `artists` that have `axes`.